lib/src/custom_rule_generator.dart

```diff
@@ -6,43 +6,190 @@
 ///
 /// If [listFormat] is true, then the output of the rule will be a list
 /// of field values. If it is false, then the output is a map keyed by
-/// field name. The [libraryName] is used as the name of the
-/// generated library. It will import [originalImport] and expects that to
-/// contain the classes in [contents]. See the comment for transformer.dart.
-String generateCustomRulesFor(String contents,
-    {String libraryName, String originalImport, bool listFormat : true}) {
+/// field name. See the comment for transformer.dart.
+AssetSerializationAnalysisResults analyzeAsset(String contents,
+    {bool listFormat: true, processAnnotatedClasses: false,
+    processAllClasses: false}) {
 
   var lib = parseCompilationUnit(contents);
-  var classes = lib.declarations.where((x) => x is ClassDeclaration);
-  var rules =
-      classes.map((each) => new CustomRuleGenerator(each, listFormat)).toList();
-  var mapOfRuleNamesToRules = rules
-      .map((x) => "    '${x._declaration.name}' : new ${x.ruleName}()")
-      .join(",\n");
-  var ruleDeclarations = rules.map((x) => x.rule).join("\n\n");
 
-  return '''
-// Generated serialization rules. *** DO NOT EDIT ***
-// See transformer.dart in package serialization.
-library ${libraryName}_serialization_rules;
+  // Check if the file is a part of another or is the main library file and find
+  // the library name.
+  String libraryName;
+  PartOfDirective partOf = lib.directives.firstWhere(
+          (x) => x is PartOfDirective, orElse: () => null);
+  if(partOf != null) {
+    libraryName = "${partOf.libraryName}";
+  } else {
+    LibraryDirective library = lib.directives.firstWhere(
+            (x) => x is LibraryDirective, orElse: () => null);
+    libraryName = library == null ? "__default" : "${library.name}";
+  }
+
+  // Find out if the file has a "package:serialization/serialization.dart" or a
+  // "package:serialization/serialization_mirrors.dart" import statement.
+  bool importsSerialization = lib.directives.firstWhere(
+      (x) => (x is ImportDirective) &&
+             "${x.uri}".contains("package:serialization/serialization.dart"),
+      orElse: () => null) != null;
+
+  bool importsSerializationMirror = lib.directives.firstWhere(
+      (x) => (x is ImportDirective) &&
+             "${x.uri}".contains("package:serialization/"
+                                 "serialization_mirrors.dart"),
+      orElse: () => null) != null;
+
+  // Check if the classes are annotated with @Serializable
+  var classes = lib.declarations.where((x) {
+    if (x is ClassDeclaration) {
+      if (processAllClasses) return true;
+      else return processAnnotatedClasses
+          && x.metadata.any((m) => "${m.name}" == "$Serializable"
+                                   || "${m.name}" == "serializable");
+    }
+    return false;
+  });
+
+  // There was no class declaration in this file or they were not annotated with
+  // @Serializable when required.
+  if (classes.length == 0) {
```

Comment:
should be .isEmpty, but more generally, it would be better not to have to write this so that the empty list is a special case. And even more generally, I think there are better approaches than the placeholder replacement.